findatastructures/src/finmap/mod.rs

```diff
@@ -0,0 +1,28 @@
+pub trait FinMap: PartialEq + Clone {
+    /// The type of keys in the map
+    type K;
+    /// The type of values in the map
+    type V;
+
+    /// Create a new empty map
+    fn new() -> Self;
+    /// Check if the map contains a key
+    fn contains_key(&self, k: &Self::K) -> bool;
+    /// Insert a key-value pair into the map
+    fn insert(&self, k: Self::K, v: Self::V) -> Self;
```

Comment:
`insert` usually returns the value that used to be in the map at `k`. What's the design rational behind returning a copy? Also, when returning the object itself, it should consume `self`. Similarly for `remove`.

Reply:
```
   = help: message: For now, ensures clause don't work on function that have `&mut` inputs (see https://github.com/hacspec/hacspec-v2/issues/290)
```

If I understand your comment correctly, these functions ideally should mutate the struct rather than return a copy, however the design is as such due to the limitations in the ensures macro

Thus if this function were to return just the replaced value while being immutable, it is equivalent to `get`